### ISSUE TITLE
Bug: Missing data proportion recovery

### ISSUE THREAD

**Issue by seabbs** [epinowcast/epinowcast#165]
Since I am running out of possible implementation-level explanations for the observed behavior, I am starting to think about the model specification...

What seems to happen (not sure if this is the only problem) is that if specified flexibly, the share of cases with missing reference can compensate for a bad fit of `imp_obs` by producing too many cases with missing reference and distributing them over the delays. Since the likelihood for the cases with missing reference is less sharp (integrating over delays) the model can get away with that especially if case numbers are low, see the examples below.

My question is whether this is maybe an artefact of the parametric delay distribution not fitting super well to this real-world dataset. Perhaps we should do some tests with simulated data where everything is well-behaved.
I could also feed the Germany dataset with the simulated missingness into my model implementation (since it has a non-parametric reporting delay) and see what happens.

### Example 1  (40% missingness)
![02 imp_obs long](https://user-images.githubusercontent.com/16266527/187884149-c105f242-d6c1-4a3e-8031-b37b8ccecbc2.png)
![02 miss prop long](https://user-images.githubusercontent.com/16266527/187884169-cb118bba-ef25-474e-9831-41d95ad39432.png)
(note that the dashed line is the realized missing proportion through binomial sampling in the simulation)

It even looks as if the share of cases with missing reference is also capturing some of the weekday effect... This would be quite weird, and I haven't seen something like that with my models.

### Example 2 (10% missingness)
![03 imp_obs](https://user-images.githubusercontent.com/16266527/187884634-7aeef805-70e1-4999-a41a-64f993372d68.png)
![03 miss prop](https://user-images.githubusercontent.com/16266527/187884652-704404c8-2353-4eec-aef8-1ce092270224.png)
This basically illustrates that the problem does not go away even if missingness is small.

_Originally posted by @adrian-lison in https://github.com/epinowcast/epinowcast/issues/147#issuecomment-1234041066_

**Comment by seabbs** [comment]
Making use of #154 and specifying a weekly random walk on the growth rate with the day of the week effect to explore the impact of a less flexible expectation model I see the following with 20% missing data. 

It still has the same issue as you show here. It also has quite a few iterations at the maximum tree depth and is in general not well-behaved. 

<img width="735" alt="Screenshot 2022-09-01 at 20 55 50" src="https://user-images.githubusercontent.com/20317649/188001271-5f3e6917-f3cb-40a9-82f4-d1ec77353244.png">
<img width="736" alt="Screenshot 2022-09-01 at 20 55 58" src="https://user-images.githubusercontent.com/20317649/188001277-20e8257f-7703-4243-a706-75569a2f5927.png">
<img width="734" alt="Screenshot 2022-09-01 at 20 56 09" src="https://user-images.githubusercontent.com/20317649/188001282-32d56e1a-09a2-4b94-9c3e-2f26fb93f701.png">
<img width="734" alt="Screenshot 2022-09-01 at 20 56 17" src="https://user-images.githubusercontent.com/20317649/188001284-49e17d1d-7376-474a-83e8-d79d5404f814.png">
<img width="745" alt="Screenshot 2022-09-01 at 21 00 01" src="https://user-images.githubusercontent.com/20317649/188001561-4cf9d9fa-9842-4cb4-b725-17e99c0459b4.png">


## Example code

```r
# Load packages
library(epinowcast)
library(data.table)
library(purrr)
library(ggplot2)

# Use 4 cores
options(mc.cores = 4)

# Load and filter germany hospitalisations
nat_germany_hosp <- germany_covid19_hosp[location == "DE"][age_group %in% "00+"]
nat_germany_hosp <- enw_filter_report_dates(
  nat_germany_hosp,
  latest_date = "2021-09-01"
)

# Make sure observations are complete
nat_germany_hosp <- enw_complete_dates(
  nat_germany_hosp,
  by = c("location", "age_group"),
  missing_reference = FALSE
)

# Set proportion missing at 20%
prop_miss <- 0.2

# Simulate
nat_germany_hosp <- enw_simulate_missing_reference(
  nat_germany_hosp,
  proportion = prop_miss, by = c("location", "age_group")
)

# Make a retrospective dataset
retro_nat_germany <- enw_filter_report_dates(
  nat_germany_hosp,
  remove_days = 40
)
retro_nat_germany <- enw_filter_reference_dates(
  retro_nat_germany,
  include_days = 60
)

# Get latest observations for the same time period
latest_obs <- enw_latest_data(nat_germany_hosp)
latest_obs <- enw_filter_reference_dates(
  latest_obs,
  remove_days = 40, include_days = 60
)

# Preprocess observations (note this maximum delay is likely too short)
pobs <- enw_preprocess_data(retro_nat_germany, max_delay = 20)

# Compile nowcasting model without multi-threading as only using a single
# group and the missing reference only supports multi-threading across groups
model <- enw_model(threads = FALSE)

# Fit the nowcast model with support for observations with missing reference
# dates and produce a nowcast
# Note that we have reduced samples for this example to reduce runtimes
nowcast <- epinowcast(pobs,
  expectation = enw_expectation(~ rw(week) + (1 | day_of_week), data = pobs),
  missing = enw_missing(~ (1 | week), data = pobs),
  report = enw_report(~ (1 | day_of_week), data = pobs),
  fit = enw_fit_opts(
    save_warmup = FALSE, pp = TRUE,
    chains = 4, iter_warmup = 500, iter_sampling = 500,
    likelihood_aggregation = "groups", adapt_delta = 0.9
  ),
  obs = enw_obs(family = "poisson", data = pobs),
  model = model
)
```

**Comment by seabbs** [comment]
Another option for digging into this is adapting the data simulator @sbfnk wrote for this [nowcasting case study](https://github.com/epiforecasts/nowcasting.example). At some point it would be nice to have a `epinowcast` example here making use of all the new tooling so any efforts could be used for more than debugging.

**Comment by adrian-lison** [comment]
> The only in model explanation I can think of here is that there is a leak between the expectation and the observed data with either less that should be being allocated to missing or more than should be allocated to reported.

True, but I already checked for that rather extensively, and the overall expected observations by reference date are exactly distributed to the expected missing/non-missing observations by reporting date

**Comment by seabbs** [comment]
Yes I know so I don't think it can be that. 😭

**Comment by sbfnk** [comment]
One thing I tried is fitting the model with a fixed missing proportion and then re-estimating using the observations generated from the model as data. The problem persists if potentially at a slightly smaller level, suggesting to me that there is at least some issue with identifiability.

<img width="775" alt="image" src="https://user-images.githubusercontent.com/1156307/196999404-caea561f-9ffd-4ca5-9978-c5c72ba3da32.png">

Relevant code is at https://github.com/sbfnk/epinowcast/blob/ad40667c7c4324c64e4972919a6f5703c4be0eb3/inst/examples/germany_missing.R#L76

**Comment by seabbs** [comment]
Really nice. The code to extract from the fit object could be very useful as part of the package. 

It is sounding that way but so surprising given @adrian-lison hasn't found that with his version of the model.  https://github.com/epinowcast/epinowcast/issues/165#issuecomment-1234719826 suggests this isn't a problem with flexibility so I wonder what other differences there are that could be driving this? 

Something I find a little surprising is that the bias is always positive. I suppose this could be because the prior is centred around 50%? Do you get the same kind of result when the proportion of missing data is very large?

**Comment by sbfnk** [comment]
Scratch what I wrote above. Having fixed a bug in my own code it now recovers the parameter nicely.

<img width="691" alt="image" src="https://user-images.githubusercontent.com/1156307/197052141-5d064eb1-5d39-4b5b-bf2e-ec89687e54fc.png">

**Comment by seabbs** [comment]
Well, that is very reassuring! This is great news. Sounds like @adrian-lison hunch may have been correct then and it's an issue with the lack of flexibility in the delay distribution.

**Comment by adrian-lison** [comment]
Yes, this looks like good news. And neat idea to refit on the model predictions, I somehow hadn't thought of that, @sbfnk ^^

@seabbs So with "varying over time" you mean the _simulated_ missing proportion, right? And you are guessing that variation in this proportion could provide more signal to differentiate between reporting delays and missingness?

**Comment by seabbs** [comment]
Yes, that is exactly the hunch @adrian-lison though may not turn out to be the case.

**Comment by seabbs** [comment]
As we now have support for non-parametric models I think we may be able to wake this back up and check to see if that was the root cause. My read of the above (and @sbfnk showing recovering) indicates that it is verly likely to be the case.
